Code/Framework/AzToolsFramework/AzToolsFramework/ComponentMode/ComponentModeSwitcher.cpp

```diff
@@ -0,0 +1,381 @@
+            m_switcherId,
+            buttonId,
+            buttonTooltip);
+    }
+
+    void ComponentModeSwitcher::RemoveComponentButton(const AZ::EntityComponentIdPair pairId)
+    {
+        // find pairId in m_whatever, call bus to remove
+        if (auto componentDataIt = AZStd::find_if(
+                m_addedComponents.begin(),
+                m_addedComponents.end(),
+                [pairId](const ComponentData& componentInfo)
+                {
+                    return componentInfo.m_pairId == pairId;
+                });
+            componentDataIt != m_addedComponents.end())
+        {
+            ViewportUi::ViewportUiRequestBus::Event(
+                ViewportUi::DefaultViewportId,
+                &ViewportUi::ViewportUiRequestBus::Events::RemoveSwitcherButton,
+                m_switcherId,
+                componentDataIt->m_buttonId);
+
+            m_addedComponents.erase(componentDataIt);
+        }
+    }
+
+    void ComponentModeSwitcher::ActivateComponentMode(const ViewportUi::ButtonId buttonId)
+    {
+        bool inComponentMode;
+        AzToolsFramework::ComponentModeFramework::ComponentModeSystemRequestBus::BroadcastResult(
+            inComponentMode, &ComponentModeSystemRequests::InComponentMode);
+        ComponentData* component = nullptr;
+
+        // find component associated with button Id
+        if (auto componentDataIt = AZStd::find_if(
+                m_addedComponents.begin(),
+                m_addedComponents.end(),
+                [buttonId](const ComponentData& componentData)
+                {
+                    return componentData.m_buttonId == buttonId;
+                });
+            componentDataIt != m_addedComponents.end())
+        {
+            component = componentDataIt;
+        }
+
+        // if already in component mode, end current mode and switch active button to the selected component
+        if (inComponentMode)
+        {
+            AzToolsFramework::ComponentModeFramework::ComponentModeSystemRequestBus::Broadcast(
+                &ComponentModeSystemRequests::EndComponentMode);
+
+            ViewportUi::ViewportUiRequestBus::Event(
+                ViewportUi::DefaultViewportId,
+                &ViewportUi::ViewportUiRequestBus::Events::SetSwitcherActiveButton,
+                m_switcherId,
+                buttonId);
+        }
+
+        // if the newly active button is not the transform button (no component mode associated), emter component mode 
+        if (buttonId != m_transformButtonId)
+        {
+            AzToolsFramework::ComponentModeFramework::ComponentModeSystemRequestBus::Broadcast(
+                &ComponentModeSystemRequests::AddSelectedComponentModesOfType, component->m_component->GetUnderlyingComponentType());
+        }
+    }
+
+    void ComponentModeSwitcher::OnEditorModeDeactivated(
+        [[maybe_unused]] const ViewportEditorModesInterface& editorModeState, [[maybe_unused]] ViewportEditorMode mode)
```

Comment:
nit: can remove `[[maybe_unused]]` from `ViewportEditorMode` as we actually are using it below